src/main/java/org/embulk/output/azure_blob_storage/BlockBlobFileOutput.java

```diff
@@ -0,0 +1,248 @@
+                    .withInitialRetryWait(500)
+                    .withMaxRetryWait(30 * 1000)
+                    .runInterruptible(new RetryExecutor.Retryable<Void>() {
+                        @Override
+                        public Void call() throws IOException, StorageException
+                        {
+                            String blockId = Base64.getEncoder().encodeToString(String.format("%10d", blockIndex).getBytes());
+                            blockBlob.uploadBlock(blockId, new BufferedInputStream(new FileInputStream(file)), file.length());
+                            blocks.add(new BlockEntry(blockId, BlockSearchMode.UNCOMMITTED));
+                            logger.info("Uploaded block file: {}, id: {}, size ~ {}kb", file.getName(), blockId, file.length() / 1024);
+                            blockIndex++;
+                            return null;
+                        }
+
+                        @Override
+                        public boolean isRetryableException(Exception exception)
+                        {
+                            return true;
+                        }
+
+                        @Override
+                        public void onRetry(Exception exception, int retryCount, int retryLimit, int retryWait)
+                                throws RetryExecutor.RetryGiveupException
+                        {
+                            if (exception instanceof FileNotFoundException || exception instanceof URISyntaxException || exception instanceof ConfigException) {
+                                throw new RetryExecutor.RetryGiveupException(exception);
+                            }
+                            String message = String.format("Azure Blob Storage put request failed. Retrying %d/%d after %d seconds. Message: %s",
+                                    retryCount, retryLimit, retryWait / 1000, exception.getMessage());
+                            if (retryCount % 3 == 0) {
+                                logger.warn(message, exception);
+                            }
+                            else {
+                                logger.warn(message);
+                            }
+                        }
+
+                        @Override
+                        public void onGiveup(Exception firstException, Exception lastException)
+                        {
+                        }
+                    });
+        }
+        catch (RetryExecutor.RetryGiveupException ex) {
+            throw new RuntimeException(ex.getCause());
+        }
+        catch (InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+        finally {
+            if (file.exists() && !file.delete()) {
+                logger.warn("Couldn't delete local file " + file.getAbsolutePath());
+            }
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        closeCurrentFile();
+    }
+
+    @Override
+    public void abort() {}
```

Comment:
I have a concern, we have 10 blocks to upload, and 5 blocks success, after that get the error
Do the azure store 5 success blocks and we still get charged?

Reply:
* Error during upload a block will be retried.
* Blob is ready when it's committed [See](https://github.com/embulk/embulk-output-azure_blob_storage/pull/20/files#diff-c9a24c6ce42ea5827fee485d2804cd86312b6d1fcbe2012f509ad40b287d91e5R158) If it isn't committed there's nothing get changed.

Reply:
We upload big file  (10G) to azure, from the plugin will split into small files then upload & commit.
The error happened when we upload 5G and got the error, and we don't delete anything when error happens

The concern here the azure will charge money when we still let 5G data on azure?

Reply:
uncommitted blocks are garbage collected in 1 week[see](https://stackoverflow.com/questions/68662051/how-to-delete-uncommitted-blocks-in-azure-blob-storage)
Anyways, I updated the code to delete the uncommitted blob (if not existed). With the existing blob, uncommitted blocks will be left as is because we can't delete the existing blob.